[ci skip] Adding Changelog for Release 4.2.0

ndellingwood opened the pull request:
Part of Kokkos C++ Performance Portability Programming EcoSystem 4.2

ndellingwood commented:
Snipped over from #6197

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,115 @@
+- `TeamThreadRange` `parallel_scan` with return value [\#6090](https://github.com/kokkos/kokkos/pull/6090), [\#6301](https://github.com/kokkos/kokkos/pull/6301), [\#6302](https://github.com/kokkos/kokkos/pull/6302), [\#6303](https://github.com/kokkos/kokkos/pull/6303), [\#6307](https://github.com/kokkos/kokkos/pull/6307)
+- `ThreadVectorRange` `parallel_scan` with return value [\#6235](https://github.com/kokkos/kokkos/pull/6235), [\#6242](https://github.com/kokkos/kokkos/pull/6242), [\#6308](https://github.com/kokkos/kokkos/pull/6308), [\#6305](https://github.com/kokkos/kokkos/pull/6305), [\#6292](https://github.com/kokkos/kokkos/pull/6292)
+- Add team-level std algorithms [\#6200](https://github.com/kokkos/kokkos/pull/6200), [\#6205](https://github.com/kokkos/kokkos/pull/6205), [\#6207](https://github.com/kokkos/kokkos/pull/6207), [\#6207](https://github.com/kokkos/kokkos/pull/6207), [\#6208](https://github.com/kokkos/kokkos/pull/6208), [\#6209](https://github.com/kokkos/kokkos/pull/6209), [\#6210](https://github.com/kokkos/kokkos/p ...
+- Add `is_*_view` and `is_*_view_v` where missing, [\#6195](https://github.com/kokkos/kokkos/pull/6195)
```
```suggestion
- Add missing `is_*_view` traits and `is_*_view_v` helper variable templates for `DynRankView`, `DynamicView`, `OffsetView`, `ScatterView` containers [\#6195](https://github.com/kokkos/kokkos/pull/6195)
```
Also would move to "general enhancements".

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,115 @@
+#### HIP:
+- New naming convention for AMD GPU: [\#6226](https://github.com/kokkos/kokkos/pull/6266)
+- Add initial support for gfx942: [\#6358](https://github.com/kokkos/kokkos/pull/6358)
+- Add support for ROCM 5.5 and 5.6 [\#6316](https://github.com/kokkos/kokkos/pull/6316)
```
```suggestion
```
Only affecting unit tests, not worth mentioning.

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,115 @@
+- Fixed potential data race in Cuda `parallel_reduce` [\#6236](https://github.com/kokkos/kokkos/pull/6236)
+- Use `cudaMallocAsync` by default [\#6402](https://github.com/kokkos/kokkos/pull/6402)
+- Bugfix when specifying non-default device ID while launching threads after initialization [\#6299](https://github.com/kokkos/kokkos/pull/6299)
+- Deprecate `Cuda(cudaStream_t,bool)` constructor [\#6401](https://github.com/kokkos/kokkos/pull/6401)
```
```suggestion
```
Not willing to acknowledge that constructor ever existed.
Deprecation was there to be safe.  It is good enough, no need to include in the changelog.

Same comment everywhere.

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,115 @@
+#### CUDA:
+- Fixed potential data race in Cuda `parallel_reduce` [\#6236](https://github.com/kokkos/kokkos/pull/6236)
+- Use `cudaMallocAsync` by default [\#6402](https://github.com/kokkos/kokkos/pull/6402)
+- Bugfix when specifying non-default device ID while launching threads after initialization [\#6299](https://github.com/kokkos/kokkos/pull/6299)
```
```suggestion
- Bugfix for using Kokkos from a thread of execution [\#6299](https://github.com/kokkos/kokkos/pull/6299)
```

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,115 @@
+- Add support for ROCM 5.5 and 5.6 [\#6316](https://github.com/kokkos/kokkos/pull/6316)
+- Improve reduction performance [\#6229](https://github.com/kokkos/kokkos/pull/6229)
+- Deprecate `HIP(hipStream_t,bool)` constructor [\#6401](https://github.com/kokkos/kokkos/pull/6401)
+- Update print_configuration [\#6387](https://github.com/kokkos/kokkos/pull/6387)
```
```suggestion
```

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,115 @@
+- Deprecate `Cuda(cudaStream_t,bool)` constructor [\#6401](https://github.com/kokkos/kokkos/pull/6401)
+
+#### HIP:
+- New naming convention for AMD GPU: [\#6226](https://github.com/kokkos/kokkos/pull/6266)
```
Might be worth spelling out what the new names are

dalg24 reviewed (commented):
Drive by comments

ndellingwood commented:
@dalg24 thanks for the comments and feedback, I addressed in [2753f18](https://github.com/kokkos/kokkos/pull/6583/commits/2753f18c1e58bd4af386533ac5c74b48f1759ca5) , let me know if you'd like any more refinements

Rombur commented on CHANGELOG.md:
```diff
@@ -1,6 +1,111 @@
+- Bugfix for using Kokkos from a thread of execution [\#6299](https://github.com/kokkos/kokkos/pull/6299)
+
+#### HIP:
+- New naming convention for AMD GPU: Vega906,Vega908,Vega90A,Navi1030 to Gfx906,Gfx908,Gfx90A,Gfx1030 [\#6226](https://github.com/kokkos/kokkos/pull/6266)
```
```suggestion
- New naming convention for AMD GPU: VEGA906, VEGA908, VEGA90A, NAVI1030 to AMD_GFX906, AMD_GFX908, AMD_GFX90A, AMD_GFX1030 [\#6226](https://github.com/kokkos/kokkos/pull/6266)
```

ndellingwood replied:
Thanks @Rombur , change committed

masterleinad commented on CHANGELOG.md:
```diff
@@ -1,6 +1,111 @@
+- Rename AMD GPU Architectures to align with AMD preferences: [\#6266](https://github.com/kokkos/kokkos/pull/6266)
+- Remove calling `tribits_exclude_autotools_files()` [\#6247](https://github.com/kokkos/kokkos/pull/6247)
+- Disable default oneDPL support in Trilinos [\#6342](https://github.com/kokkos/kokkos/pull/6342)
+- Only set KOKKOS_ARCH_AMD_GPU if a AMD GPU architecture is enabled [\#6365](https://github.com/kokkos/kokkos/pull/6365)
```
```suggestion
```
The problem was introduced when renaming architectures.

masterleinad commented on CHANGELOG.md:
```diff
@@ -1,6 +1,111 @@
+- Check for overflow during backend initialization [\#6159](https://github.com/kokkos/kokkos/pull/6159)
+- Reorganized `sort` headers for readability and separation of public/impl [\#6230](https://github.com/kokkos/kokkos/pull/6230)
+- Make constraints on `Kokkos::sort` more visible [\#6234](https://github.com/kokkos/kokkos/pull/6234) and cleanup API [\#6239](https://github.com/kokkos/kokkos/pull/6239)
+- `Kokkos::complex` only gets instantiated for cv-unqualified floating point types [\#6251](https://github.com/kokkos/kokkos/pull/6251)
```
```suggestion
```
Duplicate

masterleinad commented on CHANGELOG.md:
```diff
@@ -1,6 +1,111 @@
+ - Fix for corner case in `is_partitioned` [\#6257](https://github.com/kokkos/kokkos/pull/6257)
+ - Fix initialization of scratch lock variables in the `Cuda` backend [\#6433](https://github.com/kokkos/kokkos/pull/6433)
+ - Fixes for `Kokkos::Array` [\#6372](https://github.com/kokkos/kokkos/pull/6372)
+ - Fixes for LLVM-Clacc for OpenAPP [\#6240](https://github.com/kokkos/kokkos/pull/6240) and [\#6250](https://github.com/kokkos/kokkos/pull/6250)
```
```suggestion
 - Fixes for LLVM-Clacc for OpenACC [\#6240](https://github.com/kokkos/kokkos/pull/6240) and [\#6250](https://github.com/kokkos/kokkos/pull/6250)
```

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,111 @@
+- Bugfix for using Kokkos from a thread of execution [\#6299](https://github.com/kokkos/kokkos/pull/6299)
+
+#### HIP:
+- New naming convention for AMD GPU: Vega906,Vega908,Vega90A,Navi1030 to Gfx906,Gfx908,Gfx90A,Gfx1030 [\#6226](https://github.com/kokkos/kokkos/pull/6266)
```
Agree this is better.  @Rombur please weigh in.

ndellingwood commented:
@masterleinad I pushed the suggested changes in [9db9e1a](https://github.com/kokkos/kokkos/pull/6583/commits/9db9e1a718f1242d70ed60a93d1129ae2029779b)

masterleinad commented on CHANGELOG.md:
```diff
@@ -1,6 +1,111 @@
+
+- Export `Kokkos_CXX_COMPILER_VERSION` [\#6282](https://github.com/kokkos/kokkos/pull/6282)
+- Rename AMD GPU Architectures to align with AMD preferences: [\#6266](https://github.com/kokkos/kokkos/pull/6266)
+- Remove calling `tribits_exclude_autotools_files()` [\#6247](https://github.com/kokkos/kokkos/pull/6247)
```
What about this one? It's purely internal.

masterleinad commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+- Bugfix for using Kokkos from a thread of execution [\#6299](https://github.com/kokkos/kokkos/pull/6299)
+
+#### HIP:
+- New naming convention for AMD GPU: VEGA906, VEGA908, VEGA90A, NAVI1030 to AMD_GFX906, AMD_GFX908, AMD_GFX90A, AMD_GFX1030 [\#6226](https://github.com/kokkos/kokkos/pull/6266)
```
```suggestion
- New naming convention for AMD GPU: VEGA906, VEGA908, VEGA90A, NAVI1030 to AMD_GFX906, AMD_GFX908, AMD_GFX90A, AMD_GFX1030 [\#6266](https://github.com/kokkos/kokkos/pull/6266)
```

masterleinad commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+### Build System Changes 
+
+- Export `Kokkos_CXX_COMPILER_VERSION` [\#6282](https://github.com/kokkos/kokkos/pull/6282)
+- Rename AMD GPU Architectures to align with AMD preferences: [\#6266](https://github.com/kokkos/kokkos/pull/6266)
```
```suggestion
```
Duplicate so we should either list here or under HIP but both both.

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+ - OpenMPTarget init-join fix [\#6444](https://github.com/kokkos/kokkos/pull/6444)
+ - Fix atomic operations bug for Min and Max [\#6435](https://github.com/kokkos/kokkos/pull/6435)
+ - Fix implementation for `cyl_bessel_i0` [\#6484](https://github.com/kokkos/kokkos/pull/6484)
+ - Fix various NVCC warnings [\#6483](https://github.com/kokkos/kokkos/pull/6483)
```
```suggestion
 - Fix various NVCC warnings in `BinSort` [\#6483](https://github.com/kokkos/kokkos/pull/6483)
```

masterleinad replied:
The pull request touches 5 files, so we have `BinSort`, `algorithms` tests, `Array`, `BitManipulation`, and `MathematicalFunctiobns` tests.

dalg24 replied:
These unit tests fixes are not worth mentioning in the changelog

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+ - Fixed symlink configure issue for Windows [\#6241](https://github.com/kokkos/kokkos/pull/6241)
+ - Fix cudaAPI wrapper errors for `CUDA_MALLOC_ASYNC` [\#6346](https://github.com/kokkos/kokkos/pull/6346)
+ - simd: added a missing generator ctor for avx2 float simd_mask [\#6397](https://github.com/kokkos/kokkos/pull/6397)
+ - Use execution space instance argument to get device properties in block size deduction [\#6419](https://github.com/kokkos/kokkos/pull/6419)
```
```suggestion
```
IMO not worth mentioning.  I was fixing a future bug in a world where we support multiple GPUs from the same process.

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+ - Fixes for LLVM-Clacc for OpenACC [\#6240](https://github.com/kokkos/kokkos/pull/6240) and [\#6250](https://github.com/kokkos/kokkos/pull/6250)
+ - Fixed symlink configure issue for Windows [\#6241](https://github.com/kokkos/kokkos/pull/6241)
+ - Fix cudaAPI wrapper errors for `CUDA_MALLOC_ASYNC` [\#6346](https://github.com/kokkos/kokkos/pull/6346)
+ - simd: added a missing generator ctor for avx2 float simd_mask [\#6397](https://github.com/kokkos/kokkos/pull/6397)
```
Tempted to just cast that under general SIMD improvements.

ldh4 replied:
I think this is line redundant and could be removed. This was a follow up task to fill in few missing generator constructors from a PR that newly added in more gen constructors. And there already is a line about it `- add generator constructors [\#6347](https://github.com/kokkos/kokkos/pull/6347)`.

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+ - Fixes for `Kokkos::Array` [\#6372](https://github.com/kokkos/kokkos/pull/6372)
+ - Fixes for LLVM-Clacc for OpenACC [\#6240](https://github.com/kokkos/kokkos/pull/6240) and [\#6250](https://github.com/kokkos/kokkos/pull/6250)
+ - Fixed symlink configure issue for Windows [\#6241](https://github.com/kokkos/kokkos/pull/6241)
+ - Fix cudaAPI wrapper errors for `CUDA_MALLOC_ASYNC` [\#6346](https://github.com/kokkos/kokkos/pull/6346)
```
```suggestion
```
Bug was introduced in that release cycle (i.e. not in 4.1).  There is no reason to tell users about it.

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+- All host allocation mechanisms except for `STD_MALLOC` have been deprecated [\#6341](https://github.com/kokkos/kokkos/pull/6341)
+
+### Bug Fixes
+ - Fix AVX2 SIMD support for ZEN2 AMD CPU [\#6238](https://github.com/kokkos/kokkos/pull/6238)
```
Same comment about maybe just casting that under various SIMD improvements.

ldh4 replied:
I agree with this. I think we can move this line to be right after 
`- Kokkos_ENABLE_NATIVE=ON autodetects SIMD types supported [\#6188](https://github.com/kokkos/kokkos/pull/6188)` in the simd section.

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+### Bug Fixes
+ - Fix AVX2 SIMD support for ZEN2 AMD CPU [\#6238](https://github.com/kokkos/kokkos/pull/6238)
+ - Missing memory fence in `RandomPool::free_state` functions [\#6290](https://github.com/kokkos/kokkos/pull/6290)
+ - Fix for corner case in `is_partitioned` [\#6257](https://github.com/kokkos/kokkos/pull/6257)
```
```suggestion
 - Fix for corner case in `is_partitioned` algorithm [\#6257](https://github.com/kokkos/kokkos/pull/6257)
```

fnrizzi replied:
```suggestion
 - Fix for corner case in `Kokkos::Experimental::is_partitioned` algorithm [\#6257](https://github.com/kokkos/kokkos/pull/6257)
```

fnrizzi commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+- `half_t` and `bhalf_t` mathematical functions [\#6124](https://github.com/kokkos/kokkos/pull/6124)
+- `TeamThreadRange` `parallel_scan` with return value [\#6090](https://github.com/kokkos/kokkos/pull/6090), [\#6301](https://github.com/kokkos/kokkos/pull/6301), [\#6302](https://github.com/kokkos/kokkos/pull/6302), [\#6303](https://github.com/kokkos/kokkos/pull/6303), [\#6307](https://github.com/kokkos/kokkos/pull/6307)
+- `ThreadVectorRange` `parallel_scan` with return value [\#6235](https://github.com/kokkos/kokkos/pull/6235), [\#6242](https://github.com/kokkos/kokkos/pull/6242), [\#6308](https://github.com/kokkos/kokkos/pull/6308), [\#6305](https://github.com/kokkos/kokkos/pull/6305), [\#6292](https://github.com/kokkos/kokkos/pull/6292)
+- Add team-level std algorithms [\#6200](https://github.com/kokkos/kokkos/pull/6200), [\#6205](https://github.com/kokkos/kokkos/pull/6205), [\#6207](https://github.com/kokkos/kokkos/pull/6207), [\#6207](https://github.com/kokkos/kokkos/pull/6207), [\#6208](https://github.com/kokkos/kokkos/pull/6208), [\#6209](https://github.com/kokkos/kokkos/pull/6209), [\#6210](https://github.com/kokkos/kokkos/p ...
```
6206 is repeated twice

fnrizzi replied:
6207 is duplicated

fnrizzi commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+- Fix potential data race in HIP `parallel_reduce` [\#6429](https://github.com/kokkos/kokkos/pull/6429)
+
+#### SYCL:
+- Enforce external sycl::queues to be in-order [\#6246](https://github.com/kokkos/kokkos/pull/6246)
```
```suggestion
- Enforce external `sycl::queues` to be in-order [\#6246](https://github.com/kokkos/kokkos/pull/6246)
```

fnrizzi commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+- Add missing `is_*_view` traits and `is_*_view_v` helper variable templates for `DynRankView`, `DynamicView`, `OffsetView`, `ScatterView` containers [\#6195](https://github.com/kokkos/kokkos/pull/6195)
+- Make `nvcc_wrapper` and `compiler_launcher` scripts more portable by switching to a `#!/usr/bin/env` shebang [\#6357](https://github.com/kokkos/kokkos/pull/6357)
+- Add an improved `Kokkos::malloc` / `Kokkos::free` performance test [\#6377](https://github.com/kokkos/kokkos/pull/6377)
+- Ensure `Views` with `size==0` can be used with `deep_copy` [\#6273](https://github.com/kokkos/kokkos/pull/6273))
```
extra paren at the end of the line

fnrizzi commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+
+### Features:
+- SIMD: significant improvements to SIMD support and alignment with C++26 SIMD
+  - add `abs` overload for SIMD types [\#6069](https://github.com/kokkos/kokkos/pull/6069)
```
```suggestion
  - add `Kokkos::abs` overload for SIMD types [\#6069](https://github.com/kokkos/kokkos/pull/6069)
```

fnrizzi commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+  - define simd math function overloads in the Kokkos namespace [\#6465](https://github.com/kokkos/kokkos/pull/6465), [\#6487](https://github.com/kokkos/kokkos/pull/6487)
+  - `Kokkos_ENABLE_NATIVE=ON` autodetects SIMD types supported [\#6188](https://github.com/kokkos/kokkos/pull/6188)
+- Kokkos::printf [\#6083](https://github.com/kokkos/kokkos/pull/6083)
+- `sort`: support custom comparator [\#6253](https://github.com/kokkos/kokkos/pull/6253)
```
```suggestion
- `Kokkos::sort`: support custom comparator [\#6253](https://github.com/kokkos/kokkos/pull/6253)
```

fnrizzi commented on CHANGELOG.md:
```diff
@@ -1,6 +1,104 @@
+- `KOKKOS_ASSERT`, `KOKKOS_EXPECTS`, `KOKKOS_ENSURES` are moved to header `Kokkos_Assert.hpp` [\#6445](https://github.com/kokkos/kokkos/pull/6445)
+- Add a permuted-index mode to the gups benchmark [\#6378](https://github.com/kokkos/kokkos/pull/6378)
+- Check for overflow during backend initialization [\#6159](https://github.com/kokkos/kokkos/pull/6159)
+- Reorganized `sort` headers for readability and separation of public/impl [\#6230](https://github.com/kokkos/kokkos/pull/6230)
```
```suggestion
- Reorganized sorting headers for readability and separation of public/impl [\#6230](https://github.com/kokkos/kokkos/pull/6230)
```

fnrizzi commented:
I agree with Damien , i think the simd section is substantial enough that we could just put all related PRs under there

fnrizzi reviewed (changes requested):
please see my comments , no major ones overall

ndellingwood commented:
@dalg24 @masterleinad @fnrizzi @ldh4 I'm pushing up another round of updates addressing your comments, thanks for detailed read. Once edits are complete I'll rebase down to a single commit for this PR

fnrizzi reviewed (approved):
good to me!

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,99 @@
+ - OpenMPTarget init-join fix [\#6444](https://github.com/kokkos/kokkos/pull/6444)
+ - Fix atomic operations bug for Min and Max [\#6435](https://github.com/kokkos/kokkos/pull/6435)
+ - Fix implementation for `cyl_bessel_i0` [\#6484](https://github.com/kokkos/kokkos/pull/6484)
+ - Fix various NVCC warnings in `BinSort`, `algorithms` tests, `Array`, `BitManipulation`, and `MathematicalFunctions` tests [\#6483](https://github.com/kokkos/kokkos/pull/6483)
```
```suggestion
 - Fix various NVCC warnings in `BinSort`, `Array`, and bit manipulation function templates [\#6483](https://github.com/kokkos/kokkos/pull/6483)
```

masterleinad commented on CHANGELOG.md:
```diff
@@ -1,6 +1,99 @@
+- `KOKKOS_ASSERT`, `KOKKOS_EXPECTS`, `KOKKOS_ENSURES` are moved to header `Kokkos_Assert.hpp` [\#6445](https://github.com/kokkos/kokkos/pull/6445)
+- Add a permuted-index mode to the gups benchmark [\#6378](https://github.com/kokkos/kokkos/pull/6378)
+- Check for overflow during backend initialization [\#6159](https://github.com/kokkos/kokkos/pull/6159)
+- Reorganized headers for sorting API for readability and separation of public/impl [\#6230](https://github.com/kokkos/kokkos/pull/6230)
```
Not quite sure if this user-facing.

dalg24 replied:
You're right.  Drop it.

dalg24 replied:
```suggestion
```

masterleinad reviewed (approved):
Looks good enough.

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,99 @@
+ - Fix for corner case in `Kokkos::Experimental::is_partitioned` algorithm [\#6257](https://github.com/kokkos/kokkos/pull/6257)
+ - Fix initialization of scratch lock variables in the `Cuda` backend [\#6433](https://github.com/kokkos/kokkos/pull/6433)
+ - Fixes for `Kokkos::Array` [\#6372](https://github.com/kokkos/kokkos/pull/6372)
+ - Fixes for LLVM-Clacc for OpenACC [\#6240](https://github.com/kokkos/kokkos/pull/6240) and [\#6250](https://github.com/kokkos/kokkos/pull/6250)
```
Agree.

dalg24 commented on CHANGELOG.md:
```diff
@@ -1,6 +1,99 @@
+ - Fix for corner case in `Kokkos::Experimental::is_partitioned` algorithm [\#6257](https://github.com/kokkos/kokkos/pull/6257)
+ - Fix initialization of scratch lock variables in the `Cuda` backend [\#6433](https://github.com/kokkos/kokkos/pull/6433)
+ - Fixes for `Kokkos::Array` [\#6372](https://github.com/kokkos/kokkos/pull/6372)
+ - Fixes for LLVM-Clacc for OpenACC [\#6240](https://github.com/kokkos/kokkos/pull/6240) and [\#6250](https://github.com/kokkos/kokkos/pull/6250)
```
```suggestion
```

dalg24 commented:
Thank you all
